File: DataStructure/src/BinTree.cpp
/*链式结构存储的二叉树*/

#include "BinTree.h"
#include "SqStack.h"
#include "fatal.h"
#include "stdio.h"
#include <iostream>
#include <string>
#include "string.h"
using namespace std;

//Creat a linked storage structure of binary tree with preorder
BinTree Create_BinTree() {
	//The legal inputs are numerical string and "#"("#" represents empty node)
	BinTree L;
	string e;
	cin >> e;
	if(e == "#") {
		L = NULL;
	}
	else {
		L = new BinNode;
		L->data = e;
		L->lchild = Create_BinTree();
		L->rchild = Create_BinTree();
	}
	return L;
}

//Print the node being visited now
void Visit(BinTree T) {
	cout << T->data << " ";
}

//Convert numerical string into int value
int ToInt(BinTree T){
	char buf[12];
	int value;
	strcpy(buf,T->data.c_str());
	sscanf(buf,"%d",&value);
	return value;
}

//Recusively preorder traverse binary tree
void PreOrder(BinTree T) {
	if(T != NULL){
		Visit(T);
		PreOrder(T->lchild);
		PreOrder(T->rchild);
	}
}

//Recusively inorder traverse binary tree
void InOrder(BinTree T) {
	if(T != NULL){
		InOrder(T->lchild);
		Visit(T);
		InOrder(T->rchild);
	}
}

//Recusively postorder traverse binary tree
void PostOrder(BinTree T) {
	if(T != NULL){
		PostOrder(T->lchild);
		PostOrder(T->rchild);
		Visit(T);
	}
}

int Depth(BinTree T) {
	if(T == NULL) return 0;
	else {
		int ldepth,rdepth;
		return (ldepth = Depth(T->lchild)) > (rdepth = Depth(T->rchild)) ? (ldepth + 1) : (rdepth + 1);
	}
}

File: tree/bintree/bintree/BinTree.go
/**
* Title: Binary Tree abstract Datasture
* Time: 2017/3/24
* Author: Lee
**/

package bintree

import (
	"bufio"
	"fmt"
	"os"
	op "tree/bintree/operate"
)

type BinTree struct {
	Lchild, Rchild *BinTree
	Data           string
}

var Stdin = bufio.NewReader(os.Stdin)

func (t *BinTree) CreateBinTree() (v *BinTree) {
	/*
	** Recusively create a binary tree.The legal inputs are numerical string or "#"("#" represents empty node). The input sequence is as follows:
	** A binary tree is :
	** ----1----
	** --2---3--
	** the input is : 1 2 # # 3 # #
	 */
	var e string
	fmt.Fscanf(Stdin, "%s", &e)
	Stdin.ReadString('\n') // clean stdin buffer
	if e == "#" {
		t = nil
	} else {
		t = new(BinTree)
		t.Data = e
		t.Lchild = t.CreateBinTree()
		t.Rchild = t.CreateBinTree()
	}
	v = t
	return
}

func (t *BinTree) Value() string {
	return t.Data
}

func (t *BinTree) Left() op.Interface {
	return t.Lchild
}

func (t *BinTree) Right() op.Interface {
	return t.Rchild
}

func (t *BinTree) IsNil() bool {
	return t == nil
}
